### PR TITLE
Dev.recsys

### PR DESCRIPTION
**Pull request by torchme**
Путь к файлу notebooks/recsys.ipynb

Выполнил все что требовалось в задании

### Функция принимаеь на вход:
- Словарь с инициализированными моделями
- Словарь с инициализированными метриками
- Инициализированный Splitter для кросс-валидации
- Количество рекомендаций для генерации (K)

```
K = 10

n_splits = 3

splitter = TimeRangeSplitter(
    test_size="7D",
    n_splits=n_splits,
    filter_already_seen=True,
    filter_cold_items=True,
    filter_cold_users=True,
)

models = {
    "random": RandomModel(random_state=42),
    "popular": PopularModel(),
}

metrics = {
    "MAPTop-1": MAP(k=1),
    "MAPTop-5": MAP(k=5),
    "MAPTop-10": MAP(k=10),
    "NDCG@1": NDCG(k=1),
    "NDCG@5": NDCG(k=5),
    "NDCG@10": NDCG(k=10),
    "Precision@1": Precision(k=1),
    "Precision@5": Precision(k=5),
    "Precision@10": Precision(k=10),
    "Recall@1": Recall(k=1),
    "Recall@5": Recall(k=5),
    "Recall@10": Recall(k=10),
    "Novelty@1": MeanInvUserFreq(k=1),
    "Novelty@5": MeanInvUserFreq(k=5),
    "Novelty@10": MeanInvUserFreq(k=10),
    "Serendipity": Serendipity(k=1),
    "Serendipity": Serendipity(k=5),
    "Serendipity": Serendipity(k=10),
}
```

### Реализация обучения и валидации:
- Создаем RecTools Dataset через метод construct на train взаимодействиях для каждого фолда
- Обучаем модель (не забываем сделать deepcopy), рекоменуем K айтемов для каждого юзера, считаем метрики на test
- Дополнительно логируем время обучения
- Сохраняем метрики в отчёт
Результат оборачиваем в pandas DataFrame и усредняем по фолдам

```
import time
import pickle
from tqdm import tqdm

class RecSysModule:
    def __init__(
        self, model_dict: dict = None, metrics_dict: dict = None, splitter: TimeRangeSplitter = None, k: int = 10
    ):
        """
        Initialize the RecSysModule object.

        Parameters
        ----------
        model_dict : dict, optional
            Dictionary containing models, by default None
        metrics_dict : dict, optional
            Dictionary containing metrics, by default None
        splitter : TimeRangeSplitter, optional
            TimeRangeSplitter object for splitting data, by default None
        k : int, optional
            Value of 'k' for top-k recommendations, by default 10

        Raises
        ------
        ValueError
            If splitter is not provided.
        ValueError
            If metrics_dict is not provided.
        ValueError
            If model_dict is not provided.
        ValueError
            If no metrics are provided.
        ValueError
            If no models are provided.
        ValueError
            If k is negative.
        """
        if splitter is None:
            raise ValueError("Please provide splitter")
        if metrics_dict is None:
            raise ValueError("Please provide metrics_dict")
        if model_dict is None:
            raise ValueError("Please provide model_dict")
        if len(metrics_dict) == 0:
            raise ValueError("Please provide at least one metric")
        if len(model_dict) == 0:
            raise ValueError("Please provide at least one model")
        if k < 0:
            raise ValueError("k cannot be negative")

        self.model_dict = model_dict
        self.metrics_dict = metrics_dict
        self.splitter = splitter
        self.k = k

    def evaluate(self, interaction: pd.DataFrame) -> pd.DataFrame:
        results = []

        fold_iterator = self.splitter.split(interaction, collect_fold_stats=True)
        if len(self.splitter.get_test_fold_borders(interaction)) < 1:
            raise ValueError("Please provide at least one test fold")
        
        for train_ids, test_ids, fold_info in tqdm((fold_iterator), total=n_splits):
            start_time = time.time()

            train = interaction.df.iloc[train_ids]
            test = interaction.df.iloc[test_ids][Columns.UserItem]

            dataset = Dataset.construct(train)
            test_users = np.unique(test[Columns.User])
            catalog = train[Columns.Item].unique()

            for model_name, model in models.items():
                model.fit(dataset)
                
                with open(f'../models/{model_name}_{fold_info["i_split"]}.pickle', 'wb') as f:
                    pickle.dump(model, f)

                recos = model.recommend(
                    users=test_users,
                    dataset=dataset,
                    k=self.k,
                    filter_viewed=True,
                )
                metric_values = calc_metrics(
                    metrics,
                    reco=recos,
                    interactions=test,
                    prev_interactions=train,
                    catalog=catalog,
                )
                end_time = time.time()
                evaluation_time = end_time - start_time

                res = {"fold": fold_info["i_split"], "model": model_name, "evaluation_time": evaluation_time}
                res.update(metric_values)
                results.append(res)

        self.metrics_result = results

    def visualize_results(self):
            pivot_results = pd.DataFrame(self.metrics_result).drop(columns="fold").groupby(["model"], sort=False).agg(["mean", "std"])
            mean_metric_subset = [(metric, agg) for metric, agg in pivot_results.columns if agg == 'mean']
            return (
                pivot_results.style
                .highlight_min(subset=mean_metric_subset, color='lightcoral', axis=0)
                .highlight_max(subset=mean_metric_subset, color='lightgreen', axis=0)
            )

```

### Функция будет принимать на вход:
- Инициализированную (и обученную!) модель
- Датасет
- Список отобранных user_id для просмотра
- item_data - данные об айтемах, которые важно отразить для визуального анализа (напр. название)

```
with open('../models/popular_0.pickle', 'rb') as f:
    random_model = pickle.load(f)

users2test = [666262, 672861, 955527]

dataset = Dataset.construct(inter)

recsys.visualize_analysis(
    random_model,
    dataset=dataset,
    user_ids=users2test,
    item_data=items,
)
```

### Реализация:
- Генерим рекомендации для отобранных user_id
- Получаем из датасета истории просмотров юзеров
- Отображаем в любом удобном виде, который позволит смотреть на историю просмотра юзера + на его рекомендации от модели

```
class RecSysModule:
   ... 
   def visualize_analysis(self, model, dataset, user_ids, item_data):
        recos = model.recommend(users=user_ids, dataset=dataset, k=self.k, filter_viewed=True)
        recos = recos[recos[Columns.User].isin(user_ids)]
        recos = recos.merge(item_data, on=[Columns.Item], how='left')[['title', 'genres', Columns.User, Columns.Item]]
        
        history = dataset.interactions.df
        history = history[history[Columns.User].isin(user_ids)]
        history = history.merge(item_data, on=[Columns.Item], how='left')[['title', 'genres', Columns.User, Columns.Item]]

        for user in user_ids:
            print(f"User {user} history:")
            display(history[history[Columns.User] == user])
            print()
            print(f"User {user} recommendations:")
            display(recos[recos[Columns.User] == user])
            print()
            print()
```

### Модели:
- rectools.models.RandomModel(random_state=32)
- rectools.models.PopularModel() с параметрами по умолчанию
### Метрики: 
-  2 ранжирующие, 2 классификационные, 2 beyond-accuracy. 
- Считаем по порогам 1, 5, 10. MAP обязательно
- Сплиттер: rectools.model_selection.TimeRangeSplitter, 3 фолда для кросс-валидации по неделе, исключение холодных юзеров и айтемов и просмотренных айтемов

- Визуализация рекомендаций и историй просмотров для юзеров [666262, 672861, 955527]. 
- Для айтемов обязательно отражаем названия, жанры и количество просмотров в датасете (как для айтемов из истории взаимодействий каждого юзера, так и для айтемов из его рекомендаций)

### CODE REVIEW

**Comment by irsafilo** [comment]
все верно, но во вторую неделю поэтому 7 баллов
